### PR TITLE
transforms: (canonicalize) Add full dce

### PR DESCRIPTION
**Pull request by alexarice**
- Adds ability to specify arbitrary action after processing the worklist in the rewriter
- Canonicalize now adds full dce passes between applying rounds of canonicalization patterns
- Lets the full dce algorithm emit listener events (I'm not sure if this was necessary)
- Updates riscv canonicalization test that now gets more aggressively simplified
- Adds `cf` filecheck test demonstrating the effect of this change

### CODE REVIEW

File: tests/filecheck/dialects/riscv_cf/canonicalize.mlir
Comment (superlopuh) [inline tests/filecheck/dialects/riscv_cf/canonicalize.mlir:1]: nice

**Review by superlopuh — APPROVED** [review]

File: tests/filecheck/dialects/riscv_cf/canonicalize.mlir
Comment (alexarice) [reply tests/filecheck/dialects/riscv_cf/canonicalize.mlir:1]: Just saw this deleted the first `riscv.label` as it appears in a dead block. Is this acceptable?

Comment (superlopuh) [reply tests/filecheck/dialects/riscv_cf/canonicalize.mlir:1]: Yep! The only label we care about is the one at the start of the block we jump to

File: xdsl/pattern_rewriter.py [hunk 1]
```diff
@@ -668,6 +668,13 @@ class PatternRewriteWalker:
     That way, all uses are replaced before the definitions.
     """
 
+    post_walk_pass: Callable[[Operation, PatternRewriterListener], bool] | None = field(
+        default=None
+    )
+    """
+    Pass to call between each walk of the IR.
```
Comment (math-fehr) [inline xdsl/pattern_rewriter.py:675]:
```suggestion
    Function to call between each walk of the IR.
```

File: xdsl/pattern_rewriter.py [hunk 2]
```diff
@@ -668,6 +668,13 @@ class PatternRewriteWalker:
     That way, all uses are replaced before the definitions.
     """
 
+    post_walk_pass: Callable[[Operation, PatternRewriterListener], bool] | None = field(
```
Comment (math-fehr) [inline xdsl/pattern_rewriter.py:671]:
```suggestion
    post_walk_func: Callable[[Operation, PatternRewriterListener], bool] | None = field(
```

File: xdsl/pattern_rewriter.py [hunk 3]
```diff
@@ -755,6 +762,8 @@ def rewrite_op(self, op: Operation) -> bool:
 
         self._populate_worklist(op)
         op_was_modified = self._process_worklist(pattern_listener)
+        if self.post_walk_pass is not None:
+            op_was_modified |= self.post_walk_pass(op, pattern_listener)
```
Comment (math-fehr) [inline xdsl/pattern_rewriter.py:766]:
```suggestion
        if self.post_walk_func is not None:
            op_was_modified |= self.post_walk_func(op, pattern_listener)
```

File: xdsl/pattern_rewriter.py [hunk 4]
```diff
@@ -764,6 +773,8 @@ def rewrite_op(self, op: Operation) -> bool:
         while op_was_modified:
             self._populate_worklist(op)
             op_was_modified = self._process_worklist(pattern_listener)
+            if self.post_walk_pass is not None:
+                op_was_modified |= self.post_walk_pass(op, pattern_listener)
```
Comment (math-fehr) [inline xdsl/pattern_rewriter.py:777]:
```suggestion
            if self.post_walk_func is not None:
                op_was_modified |= self.post_walk_func(op, pattern_listener)
```

File: xdsl/transforms/canonicalize.py [hunk 5]
```diff
@@ -37,4 +37,4 @@ def apply(self, ctx: MLContext, op: builtin.ModuleOp) -> None:
         pattern = GreedyRewritePatternApplier(
             [RemoveUnusedOperations(), CanonicalizationRewritePattern()]
         )
-        PatternRewriteWalker(pattern).rewrite_module(op)
+        PatternRewriteWalker(pattern, post_walk_pass=op_dce).rewrite_module(op)
```
Comment (math-fehr) [inline xdsl/transforms/canonicalize.py:40]:
```suggestion
        PatternRewriteWalker(pattern, post_walk_func=op_dce).rewrite_module(op)
```

**Review by math-fehr — APPROVED** [review]
Nice!
I'm just proposing a different name, but otherwise good for me ✅

**Pull request merged**